File: Algorithms/0097_Interleaving_String/Python/Interleaving_String_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        length_1, length_2, length_3 = len(s1), len(s2), len(s3)
        if length_1 == length_2 == length_3 == 0: return True
        if length_1 + length_2 != length_3: return False

        cache = [[None for _ in range(length_1)] for _ in range(length_2)]

        def dfs(board, s1, s2, s3, index1, index2, index3):
            if index1 == len(s1): return s2[index2:] == s3[index3:]
            if index2 == len(s2): return s1[index1:] == s3[index3:]
            if board[index2][index1] is not None: return board[index2][index1]

            ans = False
            if s1[index1] == s3[index3] and dfs(board, s1, s2, s3, index1 + 1, index2, index3 + 1):
                ans = True
            if s2[index2] == s3[index3] and dfs(board, s1, s2, s3, index1, index2 + 1, index3 + 1):
                ans = True

            board[index2][index1] = ans
            return ans

        return dfs(cache, s1, s2, s3, 0, 0, 0)

File: Algorithms/0097_Interleaving_String/Java/InterleavingStringSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() == s2.length() && s2.length() == s3.length() && s3.length() == 0) return true;
        if (s1.length() + s2.length() != s3.length()) return false;
        int[][] cache = new int[s1.length()][s2.length()];
        for (int[] row : cache) {
            Arrays.fill(row, -1);
        }
        return dfs(cache, s1, s2, s3, 0, 0, 0);
    }

    private boolean dfs(int[][] board, String s1, String s2, String s3, int i1, int i2, int i3) {
        if (i1 == s1.length()) return s2.substring(i2).equals(s3.substring(i3));
        if (i2 == s2.length()) return s1.substring(i1).equals(s3.substring(i3));
        if (board[i1][i2] != -1) return board[i1][i2] == 1;

        boolean ans = false;
        if (s1.charAt(i1) == s3.charAt(i3) && dfs(board, s1, s2, s3, i1 + 1, i2, i3 + 1)) ans = true;
        if (s2.charAt(i2) == s3.charAt(i3) && dfs(board, s1, s2, s3, i1, i2 + 1, i3 + 1)) ans = true;

        board[i1][i2] = ans ? 1 : 0;
        return ans;
    }
}




